xde/src/xde.rs

```diff
@@ -1074,60 +1075,66 @@ unsafe extern "C" fn xde_attach(
 /// Setup underlay port atop the given link.
 fn create_underlay_port(
     link_name: String,
-    mc_name: &str,
+    // This parameter is likely to be used as part of the flows work.
+    _mc_name: &str,
 ) -> Result<xde_underlay_port, OpteError> {
-    // Grab mac handle for underlying link
-    let mh = MacHandle::open_by_link_name(&link_name).map(Arc::new).map_err(
-        |e| OpteError::System {
-            errno: EFAULT,
-            msg: format!("failed to open link {link_name} for underlay: {e}"),
-        },
-    )?;
+    let link_cstr = CString::new(link_name.as_str()).unwrap();
 
-    // Get a mac client handle as well.
-    //
-    let oflags = MacOpenFlags::NONE;
-    let mch = MacClientHandle::open(&mh, Some(mc_name), oflags, 0)
-        .map(Arc::new)
-        .map_err(|e| OpteError::System {
+    let link_id =
+        LinkId::from_name(link_cstr).map_err(|err| OpteError::System {
             errno: EFAULT,
-            msg: format!("mac_client_open failed for {link_name}: {e}"),
+            msg: format!("failed to get linkid for {link_name}: {err}"),
         })?;
 
+    let mph = MacPerimeterHandle::from_linkid(link_id).map_err(|e| {
+        OpteError::System {
+            errno: EFAULT,
+            msg: format!("failed to grab MAC perimeter for {link_name}: {e}"),
+        }
+    })?;
+
+    let link = DlsLink::hold(&mph).map_err(|e| OpteError::System {
+        errno: EFAULT,
+        msg: format!(
+            "failed to grab hold on link perimeter for {link_name}: {e}"
+        ),
+    })?;
+
+    let stream = link.open_stream(&mph).map_err(|e| OpteError::System {
+        errno: EFAULT,
+        msg: format!("failed to grab open stream for {link_name}: {e}"),
+    })?;
+
+    drop(mph);
+
     // Setup promiscuous callback to receive all packets on this link.
     //
     // We specify `MAC_PROMISC_FLAGS_NO_TX_LOOP` here to skip receiving copies
     // of outgoing packets we sent ourselves.
-    let mph = mch
-        .add_promisc(
-            mac::mac_client_promisc_type_t::MAC_CLIENT_PROMISC_ALL,
-            xde_rx,
-            mac::MAC_PROMISC_FLAGS_NO_TX_LOOP,
-        )
-        .map_err(|e| OpteError::System {
-            errno: EFAULT,
-            msg: format!("mac_promisc_add failed for {link_name}: {e}"),
-        })?;
-
-    // Set up a unicast callback. The MAC address here is a sentinel value with
-    // nothing real behind it. This is why we picked the zero value in the Oxide
-    // OUI space for virtual MACs. The reason this is being done is that illumos
-    // requires that if there is a single mac client on a link, that client must
-    // have an L2 address. This was not caught until recently, because this is
-    // only enforced as a debug assert in the kernel.
```

Comment:
This should never be the case now: being visible via (and getting a stream from) DLS implies the existence of a client on the underlying MAC. This is also the root-cause of landing in software classification, as we end up with two clients.